R/methSeg.R

```diff
@@ -88,126 +195,119 @@
+    if(class(obj)=="methylRaw" | class(obj)=="methylRawDB") {
+      obj= as(obj,"GRanges")
+      ## calculate methylation score 
+      mcols(obj)$meth=100*obj$numCs/obj$coverage
+      ## select only required mcol
+      obj = obj[,"meth"]
+    }else if (class(obj)=="methylDiff" | class(obj)=="methylDiffDB") {
+      obj = as(obj,"GRanges")
+      ## use methylation difference as score
+      obj = obj[,"meth.diff"]
+    }else if (class(obj) != "GRanges"){
+      stop("only methylRaw or methylDiff objects ", 
+           "or GRanges objects can be used in this function")
+    }
+    
+    # destrand
+    strand(obj) <- "*"
+    
+    # Run segmentation
+    seg.res = .run.fastseg(obj)
   }
   
-  if("initialization" %in% names(args.Mclust)){
-    if("subset" %in% names(args.Mclust[["initialization"]])) {
-      if(length(args.Mclust[["initialization"]][["subset"]]) < 9 ){
-        stop("too few samples, increase the size of subset.") 
+  if(cores>1){
+    
+    ## Non-tabix files  
+    if (class(obj)=="methylDiff" | class(obj)=="methylRaw" | 
+        class(obj)=="GRanges"){
+      
+      ##coerce object to granges
+      if(class(obj)=="methylRaw") {
+        obj= as(obj,"GRanges")
+        ## calculate methylation score 
+        mcols(obj)$meth=100*obj$numCs/obj$coverage
+        ## select only required mcol
+        obj = obj[,"meth"]
+      } else if(class(obj)=="methylDiff") {
+        obj = as(obj,"GRanges")
+        ## use methylation difference as score
+        obj = obj[,"meth.diff"]
+      }else if (class(obj) != "GRanges"){
+        stop("only methylRaw or methylDiff objects ", 
+             "or GRanges objects can be used in this function")
+      }
+      # destrand
+      strand(obj) <- "*"
+      
+      chrs = seqlevels(obj)
+      # Split the input object based on chromosome
+      # run methSeg on chromosomes (Could be in parallel or not)
+      seg.res.list <- mclapply(chrs, function(chr){
+        seg.res = .run.fastseg(obj[seqnames(obj)==chr])
+      }, mc.cores=cores)
+      
+      # Merge resulting segments again to whole genome.
+      # suppressWarnings -> combined objs dont have chrs in common
+      seg.res = suppressWarnings( do.call("c", seg.res.list) )
+      
+      ## Tabix files
+    } else if(class(obj)=="methylDiffDB" | class(obj)=="methylRawDB"){
+      
+      .run.fastseg.tabix = function(gr0, ...){
+        
+        # adjust colnames of input GRanges to 
+        # methylKit naming convention
+        df2getcolnames = as.data.frame(gr0[1])
+        df2getcolnames$width = NULL 
+        methylKit:::.setMethylDBNames(df2getcolnames)
+        names(mcols(gr0)) = colnames(df2getcolnames)[5:ncol(df2getcolnames)]
+        
+        if("coverage" %in% names(mcols(gr0))){
+          gr0$meth = 100*gr0$numCs/gr0$coverage
+          gr0 = gr0[,"meth"]
+        }else if("meth.diff" %in% names(mcols(gr0))){
+          gr0 = gr0[,"meth.diff"]
+        }else if (class(obj) != "GRanges"){
```

Comment:
This is unnecessary since you are already checking above for the class.

Reply:
that's actually true!